debconf module

bcoca opened the pull request:
Allows you to change or preseed debian debconf settings,  some packages won't allow installing non-interactively w/o this.

TODO: 
- add ability to take array and/or file to do mutliple operations at once
- allow for complex_args 

Signed-off-by: Brian Coca <briancoca+dev@gmail.com>

bcoca committed:
debconf package
Signed-off-by: Brian Coca <briancoca+dev@gmail.com>

srgvg commented:
+1 feature wise!

bcoca committed:
fixed typo for showing diff
Signed-off-by: Brian Coca <briancoca+dev@gmail.com>

bcoca committed:
and it was supposed to be a dict
Signed-off-by: Brian Coca <briancoca+dev@gmail.com>

bcoca committed:
corrected indentation and now handles non existing previous key (not
just changed existing)

Signed-off-by: Brian Coca <briancoca+dev@gmail.com>

bcoca committed:
did some minor refactoring on docs and fixed spelling mistakes (thanks
Mikhail)
Signed-off-by: Brian Coca <briancoca+dev@gmail.com>

bcoca committed:
corrected my grammar (again Mikhail)

Signed-off-by: Brian Coca <briancoca+dev@gmail.com>

bcoca committed:
making pylint happier

Signed-off-by: Brian Coca <briancoca+dev@gmail.com>

resmo commented:
+1

jctanner commented:
@bcoca couple of issues:
- Module parameters should not be any python keywords, or "type". Use something like value_type if you must.
- Update the last 3 lines to use the new module snippet format
- Bad indentation:

```
diff --git a/library/system/debconf b/library/system/debconf
index 7324e7d..e5b8ff0 100644
--- a/library/system/debconf
+++ b/library/system/debconf
@@ -86,10 +86,10 @@ def get_selections(module, pkg):
     if rc == 0:
         selections = {}
     for line in out.splitlines():
-            #if not line.startswith('*'): # only awnsered
-            #    continue
-            (key, value) = line.split(':')
-            selections[ key.strip('*').strip() ] = value.strip()
+        #if not line.startswith('*'): # only awnsered
+        #    continue
+        (key, value) = line.split(':')
+        selections[ key.strip('*').strip() ] = value.strip()
         return selections
     else:
         module.fail_json(msg=err)
```

kuba commented:
1. Incorrect indentation indeed, although @jctanner's patch is wrong, see @sa2ajj's https://github.com/ansible/ansible/pull/5361#commitcomment-4933376
2. Unless it's intentional, @bcoca: you could also fix spelling please: %s/awnser/answer/g

kuba commented on library/system/debconf:
```diff
@@ -0,0 +1,162 @@
+        module.fail_json(msg=err)
+
+
+def set_selection(module, pkg, question, type, value, unseen):
```
`type` is a Python [built-in function](http://docs.python.org/2/library/functions.html) and shadowing/rebinding  those is considered a bad practice, I'd recommend using a different variable name

bcoca replied:
yeah, return should be outside for, reformatting that a bit to make it
clearer ... also fixing spelling

On Wed, Feb 5, 2014 at 5:56 PM, Jakub Warmuz notifications@github.comwrote:

> In library/system/debconf:
> 
> > -    cmd = [module.get_bin_path('debconf-show', True), pkg]
> > -    rc, out, err = module.run_command(' '.join(cmd))
> >   +
> > -    if rc == 0:
> > -        selections = {}
> > -    for line in out.splitlines():
> > -            #if not line.startswith('*'): # only awnsered
> > -            #    continue
> > -            (key, value) = line.split(':')
> > -            selections[ key.strip('*').strip() ] = value.strip()
> > -        return selections
> > -    else:
> > -        module.fail_json(msg=err)
> >   +
> >   +
> >   +def set_selection(module, pkg, question, type, value, unseen):
> 
> type is a Python built-in functionhttp://docs.python.org/2/library/functions.htmland shadowing/rebinding those is considered a bad practice, I'd recommend
> using a different variable name
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/5361/files#r9482729
> .

## 

Brian Coca
Stultorum infinitus est numerus
0110000101110010011001010110111000100111011101000010000001111001011011110111010100100000011100110110110101100001011100100111010000100001
Pedo mellon a minno

bcoca committed:
fixes as per feedback

jctanner commented:
@bcoca the third example also needs to make "names" singular.

bcoca committed:
changes names to name in example

jctanner commented:
@bcoca 
- main() is not at the bottom of the file, so the module isn't actually executed
- For names that have no values, the module exits with a TypeError traceback

```
diff --git a/library/system/debconf b/library/system/debconf
index 3554534..79b9259 100644
--- a/library/system/debconf
+++ b/library/system/debconf
@@ -102,7 +102,10 @@ def set_selection(module, pkg, question, vtype, value, unseen):
         answer.append(vtype)
     answer.append(value)

-    data = ' '.join(answer)
+    if answer is not None:
+        data = ' '.join(answer)
+    else:
+        #???

     setsel = module.get_bin_path('debconf-set-selections', True)
     cmd = ["echo '%s %s' |" % (pkg, data), setsel]
@@ -159,3 +162,5 @@ def main():

 # import module snippets
 from ansible.module_utils.basic import *
+
+main()
```

bcoca commented:
sorry, must have deleted it when I changed from oldstyle imports to new
style.

else:
 pass

should be used for those with no values?

## 

Brian Coca
Stultorum infinitus est numerus
0110000101110010011001010110111000100111011101000010000001111001011011110111010100100000011100110110110101100001011100100111010000100001
Pedo mellon a minno

jctanner commented:
@bcoca i'm not familiar enough with debconf to weigh in on what do with those None values. 

@kuba @resmo thoughts?

bcoca committed:
fixed main() issue created in prev fix, now values are empty string if they were None

Signed-off-by: Brian Coca <briancoca+ansible@gmail.com>

bcoca commented:
I updated it to be an empty string, which should be ok for most situations
and return a handled error if not.

kuba commented:
Per `man 1 debconf-set-selections`:

```
(...) All other lines set the value of one
question, and should contain four values, each separated by one character of
whitespace. The first value is the name of the package that owns the
question. The second is the name of the question, the third value is the type
of this question, and the fourth value (through the end of the line) is the
value to use for the answer of the question.
```

So all fields are required and cannot be left empty! It's not "ok for most situations". I'd set required flag for all: `description`, `setting`, `vtype` and `value` for now. Revoking requirement is backward compatible, whereas opposite is not true. The only problem I can see is that in the future, if you allow to pass an "array of optionas and/or debconf file from get-selections dump", you'll need to find some way of telling which options are required.

Rant about wording: I went through a few of Debian/debconf [manuals](http://manpages.debian.net/cgi-bin/man.cgi?query=debconf-set-selections), [tutorials](http://www.fifi.org/doc/debconf-doc/tutorial.html) and [policy document](https://www.debian.org/doc/packaging-manuals/debconf_specification.html)s in order to review your code. To keep things easy and not confusing to the potential users of the module or future contributors to the code, consider the following:
1. `description` is never used as an alias to the "package name". Set `pkg` as the default and optionally add `owner` as an alias.
2. Make `question` as the default. Word `setting` is never used in this context as well.
3. `vtype` is technically _question type_, so `qtype`. Although it's not used in this context anywhere else, I'd say `vtype` as an alias could be an option.
4. Change variable descriptions in `DOCUMENTATION` accordingly.
5. In `set_selections()` use `selection` instead of `answer` in order to build up the line passed to `debconf-set-selections`.
6. etc.

Btw, `-u` is interestingly not documented in the manpages, I'll consider sending a bug report upstream.

Lastly, `debconf`, although has priority set to important, is not essential package in Debian (and derivatives?), so I'd list it in the documentation as a requirement of the module.

Now a feature request. If you care to distinguish between different question types (we could just allow to set `question="debconf/priority select"`, then maybe consider the following:
1. If `qtype` is `bool` then allow `value` to be anything what currently Ansible treats as a boolean, i.e. "true", "True", "yes", etc. and cast it to legitimate "true" and "false".
2. If `qtype` is `multiselect` then allow `value` to be a list of and join items with ", ".

Sorry if I'm too picky, but I got an impression that Ansible recently tries to improve the code quality :).

bcoca commented:
You being picky makes me a better programmer ;-)

I'll look at fixing the issues this weekend, pretty busy until then. I
disagree on qtype vs vtype, its really the type of the answer as questions
are always strings.

Some things I might add into current, others might just be subsequent
patches, like allowing for multiple questions at once or even being fed
from a seed file.

## 

Brian Coca
Stultorum infinitus est numerus
0110000101110010011001010110111000100111011101000010000001111001011011110111010100100000011100110110110101100001011100100111010000100001
Pedo mellon a minno

bcoca commented:
I've left value as empty and defaulting to empty string because it is the
only way to 'unset', also I cannot make value a required parameter as it
makes no sense when using the module to query existing questions.

>    1.
> 
>    description is never used as an alias to the "package name". Set pkgas the default and optionally add
>    owner as an alias.
> 
> Not sure what you mean by this. there is 'name' and 'pkg' as an alias. I'm
> using name as the default as per ansible conventions.
> 
>    1.
> 
>    Make question as the default. Word setting is never used in this
>    context as well.
> 
> fine (still leaving setting as alias as many refer to them this way, even
> though the docs don't)
> 
>    1.
> 
>    vtype is technically _question type_, so qtype. Although it's not used
>    in this context anywhere else, I'd say vtype as an alias could be an
>    option.
> 
> in the end this is the type of the answer to the question, not the
> question itself, leaving it as it seems the most descriptive for input.
> 
>    1.
> 
>    Change variable descriptions in DOCUMENTATION accordingly.
> 
> done
> 
>    1.
> 
>    In set_selections() use selection instead of answer in order to build
>    up the line passed to debconf-set-selections.
> 
> This is an internal variable name, i think it is descriptive for the
> programmer, the user never sees it. I find it clearer than the Debian
> naming which is very confusing in this area, at least for me.
> 
> Lastly, debconf, although has priority set to important, is not essential
> package in Debian (and derivatives?), so I'd list it in the documentation
> as a requirement of the module.
> 
> is it not part of base install? I've never had to install it myself. Added
> note anyways.
> 
> Now a feature request. If you care to distinguish between different
> question types (we could just allow to set question="debconf/priority
> select", then maybe consider the following:
> 
>    1.
> 
>    If qtype is bool then allow value to be anything what currently
>    Ansible treats as a boolean, i.e. "true", "True", "yes", etc. and cast it
>    to legitimate "true" and "false".
> 
> done
> 
>    1.
> 
>    If qtype is multiselect then allow value to be a list of and join
>    items with ", ".
> 
> done

## 

Brian Coca
Stultorum infinitus est numerus
0110000101110010011001010110111000100111011101000010000001111001011011110111010100100000011100110110110101100001011100100111010000100001
Pedo mellon a minno

bcoca committed:
made changes as per feedback:

- field names are more consistent with debconf
- values are now 'booleanized' or accepted as list/set objects when
  pertinent
- updated docs to reflect all of the above and debconf cli tools
  required

kuba commented:
1. Yep, ignore my comment about `description`.
2. It wasn't obvious to me, that you can use the module to query questions... Maybe add an example? Still, to be honest I'm really confused by what is returned as `current` and `previous`. Looking at the code:
   1. If no `question` supplied, `current` is **all** debconf selections for a package.
   2. If `question` is given, but `value` is not changed then you again return **all** debconf selections.
   3. If `value` is changed and `current` is **just** the new selection for `question` and `prev` are **all** debconf selections.
   
   Is this the intended behaviour? Don't you want to set `previous` and `current` just to mention `question` queried? And if no `question` is given then maybe return all of the selections for given package? Whatever you do, you should definitely explicitly state the behaviour in comments and/or `DOCUMENTATION`.
3. How about passing to `AnsibleModule` the following?:
   
   ```
   required_together = (
       ['question', 'vtype', 'value'],
   )
   ```
   
   or
   
   ```
   required_together = (
       ['vtype', 'value'],
   )
   ```
   
   depending on what you want to achieve.
4. In general, I admit that debconf might have misleading naming conventions, but it will be even more misleading if you try to mess around them like you do now. Have a though about it; if you didn't, have a look at links to manuals and tutorials from my previous comment (https://github.com/ansible/ansible/pull/5361#issuecomment-34376788) - this will make you less confused. Mind that rebinding names used in debconf manuals is even worse, and I as a "programmer" you mentioned, had troubles deciphering the code.
   
   Thus, maybe set `qtype` as default and `vtype` as an alias? Or the other way round at least?
5. Indeed, `debconf` is part of base Debian install, but it's not marked as essential package, and could, in theory, be easily removed (I think so).

bcoca committed:
hopefully last batch of fixes

- removed previous 'typification' of input as it needs it is typed by
  module as strings and needs to be output as strings, making it
  useless.
- now checks for vtype and value against None when question is specified
- simplified set_selections as vtype and value should have a string
  value going in.
- added example of querying questions for a package
- added module requirement of question,vtype and value being required
  together.

bcoca commented:
After much testing and differnt use cases, I think I've gotten to a final submission point. 

I've made a check for None values and should return a nice error now. This had the nice side effect of simplifying other code.

I reverted the 'typification' of value as it became useless due to both input validation type and that it is converted to a string for set-selections input anyways.

I also added example of querying package questions and made the other parameters required together.

bcoca committed:
didnt make 1.5 ... lets see

jctanner commented:
@bcoca thanks!
